src/compat/libc/stdio/fread.c

```diff
@@ -13,52 +13,46 @@
 
 #include <stdio.h>
 
-size_t fread(void *buf, size_t size, size_t count, FILE *file) {
-	char *cbuff;
+static size_t __fread(FILE *file, void *buf, size_t len) {
 	size_t cnt;
 
+	if (funopen_check(file)) {
+		cnt = (file->readfn) ? file->readfn((void *) file->cookie, buf, len) : 0;
+	} else {
+		cnt = read(file->fd,  buf, len);
+	}
+	if (cnt <= 0) {
+		file->flags |= cnt ? IO_ERR_ : IO_EOF_;
+		return 0;
+	}
+	return cnt;
+}
+
+size_t fread(void *buf, size_t size, size_t count, FILE *file) {
+	char *cbuff = buf;
+	size_t len = size * count, l = len, ret;
+
 	if (NULL == file) {
 		SET_ERRNO(EBADF);
 		return -1;
 	}
 
 	if ((file->flags & O_ACCESS_MASK) == O_WRONLY) {
-		SET_IO_ERR(file);
-		return -1;
+		file->flags |= IO_ERR_;
+		return 0;
 	}
 
-	cnt = 0;
 	if (file->has_ungetc) {
 		file->has_ungetc = 0;
-		cbuff = buf;
-		cbuff[0] = (char)file->ungetc;
-		count --;
-		buf = &cbuff[1];
-		cnt++;
+		*cbuff++ = (char)file->ungetc;
+		l--;
 	}
-	while (cnt != count * size) {
-		int ret;
-		if (funopen_check(file)) {
-			if (file->readfn) {
-				ret = file->readfn((void *) file->cookie, buf, size * count);
-			} else {
-				ret = 0;
-			}
-		} else {
-			ret = read(file->fd,  buf, size * count);
-		}
-		if (ret <= 0) {
-			break; /* errors */
+	for (; l; l -= ret, cbuff += ret ) {
+		ret = (file->flags & IO_EOF_) ? 0 : __fread(file, cbuff, l);
+		if (!ret) {
+			return (len - l) / size;
```

Comment:
Just a small idea: as far I as understand fread()  executes read() two times when a stream contains less data than it was requested? If so, can we analyze the result of read and return immediately if `ret` is less than requested `l`?

Reply:
> Just a small idea: as far I as understand fread() executes read() two times when a stream contains less data than it was requested? If so, can we analyze the result of read and return immediately if `ret` is less than requested `l`?

I see what you mean.  I don't think this is quiet correct though.
AFAIU, fread() has to keep read()ing until it gets to either EOF or ERR, with them happening when read() returns 0 and -1, respectively.
As long as either one of those situations don't occur, fread() has to keep reading. This is usually a non-issue when it comes to regular files, because they tend to return as much as requested in one go. However, when it comes to other types of files, like device nodes, for instance, the amount of data returned in one read depends entirely on the driver at the other end, so it's perfectly possible to have a driver that's only able to return data in N bytes-long streams, without that being considered an error nor EOF. If we regard those cases as an implicit EOF, in my opinion, it wouldnt be compliant with POSIX.
Of course, I'm not 100% sure about this, but I think that's my current understanding of the matter.

Reply:
What if you are reading data in blocking mode? In your case, the user sees there is some data (for example, after executing select()) and attempts to read a device, read() must do not block because there is some data available - N bytes, for instance, and a user requested N + 1. On the first read() device returned N bytes, but you then request 1 additional byte in the second read() and your hangs until some new data will be available.

Reply:
Also, I am not aware of devices that return N-bytes streams only stream by stream and cannot return several N-bytes streams in one read if they are available. If such devices exist a user should be careful when reading it and knowing that it returns only a single N-bytes stream in one read and to call fread() several times from userspace.

Reply:
But your point is pretty good, and probably I am missing something,

Reply:
It looks like it's a difficult question about how generally EOF works.  For example, it is not clear how EOF should be set for pipes, I am not sure we should set EOF when reading returns 0 bytes, but when we read 0 bytes and write end is closed (I didn't check this on linux) So EOF could be probably device-dependent (but not exactly).

Reply:
Well, let's leave this part of code as is because in the previous implementation there are several potential read()'s too, so I believe nothing will be broken. I will attempt to create an issue.